sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/BaseFirestoreV1WriteFnTest.java

```diff
@@ -0,0 +1,640 @@
+    when(attempt.awaitSafeToProceed(any())).thenReturn(true);
+    when(attempt.<Write, Element<Write>>newFlushBuffer(attemptStart)).thenReturn(flushBuffer);
+    when(flushBuffer.offer(element1)).thenReturn(true);
+    when(flushBuffer.iterator()).thenReturn(newArrayList(element1).iterator());
+    when(flushBuffer.getBufferedElementsCount()).thenReturn(1);
+    when(flushBuffer.isFull()).thenReturn(true);
+
+    when(callable.call(any())).thenThrow(RETRYABLE_ERROR, RETRYABLE_ERROR, RETRYABLE_ERROR);
+    doNothing().when(attempt).recordFailedRequest(any(), anyInt());
+    doNothing().doNothing().doThrow(RETRYABLE_ERROR).when(attempt).checkCanRetry(RETRYABLE_ERROR);
+
+    when(processContext.element()).thenReturn(write);
+
+    try {
+      runFunction(getFn(clock, ff, RPC_QOS_OPTIONS));
+      fail("Expected ApiException to be throw after exhausted attempts");
+    } catch (ApiException e) {
+      assertSame(RETRYABLE_ERROR, e);
+    }
+
+    verify(attempt, times(1)).awaitSafeToProceed(attemptStart);
+    verify(attempt, times(1)).recordStartRequest(rpc1Sta);
+    verify(attempt, times(1)).recordFailedRequest(rpc1End, 1);
+    verify(attempt, times(1)).recordStartRequest(rpc2Sta);
+    verify(attempt, times(1)).recordFailedRequest(rpc2End, 1);
+    verify(attempt, times(1)).recordStartRequest(rpc3Sta);
+    verify(attempt, times(1)).recordFailedRequest(rpc3End, 1);
+    verify(attempt, times(0)).recordSuccessfulRequest(any(), anyInt());
+    verify(attempt, never()).completeSuccess();
+  }
+
+  @Override
+  @Test
+  public final void noRequestIsSentIfNotSafeToProceed() throws Exception {
+    when(ff.getFirestoreRpc(any())).thenReturn(rpc);
+    when(ff.getRpcQos(any())).thenReturn(rpcQos);
+    when(rpcQos.newWriteAttempt(FirestoreV1Fn.V1FnRpcAttemptContext.BatchWrite)).thenReturn(attempt);
+
+    InterruptedException interruptedException = new InterruptedException();
+    when(attempt.awaitSafeToProceed(any())).thenReturn(false).thenThrow(interruptedException);
+
+    when(processContext.element()).thenReturn(FirestoreProtoHelpers.newWrite());
+
+    try {
+      runFunction(getFn(clock, ff, RPC_QOS_OPTIONS));
+      fail("Expected ApiException to be throw after exhausted attempts");
+    } catch (InterruptedException e) {
+      assertSame(interruptedException, e);
+    }
+
+    verify(rpc, times(1)).close();
+    verifyNoMoreInteractions(rpc);
+    verifyNoMoreInteractions(callable);
+    verify(attempt, times(0)).recordFailedRequest(any(), anyInt());
+    verify(attempt, times(0)).recordSuccessfulRequest(any(), anyInt());
+  }
+
+  @Test
+  public abstract void enqueueingWritesValidateBytesSize() throws Exception;
+
+  @Test
+  public final void endToEnd_success() throws Exception {
+
+    Write write = FirestoreProtoHelpers.newWrite();
+    BatchWriteRequest expectedRequest = BatchWriteRequest.newBuilder()
+        .setDatabase("projects/testing-project/databases/(default)")
```

Comment:
Does this need to be parametrized on `ENV_GOOGLE_CLOUD_PROJECT`? Same on the other test cases.

Reply:
not in this case, the unit tests are all in process and use mocks for all RPCs.

Reply:
Ok.